
import csv
import os
import sys
import pytest
import logging

logger = logging.getLogger(__name__)

def read_csv(file_name):
    with open(file_name, newline='') as csvfile:
        reader = csv.reader(csvfile)
        reader = csv.DictReader(csvfile)
        next(reader)  # Skip the header row
        return list(reader)

# These reference files are generated by fixtures/downloads/download.py
# They consist of a range of variants that have been converted from HGVS to genomic coordinates by
# two third party services and are used to validate the behaviour of the in-house translator
ensembl_csv_data = read_csv('tests/fixtures/downloads/output_ensemble.csv')
variant_validator_csv_data = read_csv('tests/fixtures/downloads/output_variant_validator.csv')

# Validate each row of the ensembl file against what is produced by the in house HGVS translator
@pytest.mark.parametrize("row", ensembl_csv_data)
def test_against_ensembl(row):
    compare(row)

# Validate each row of the variant validator file against what is produced by the in house HGVS translator
@pytest.mark.parametrize("row", variant_validator_csv_data)
def test_against_variant_validator(row):
    compare(row)
    
def compare(row):
    logger.info(f"Using row: {row}")
    hgvs = row['HGVS']
    success = row['Success'] == "True"
    chromosome = row['Chromosome']
    position = row['Position']
    reference = row['Ref']
    alternate = row['Alt']
    assert chromosome == "X", f"The chromosome should be X"
    # TODO: Compare against inhouse tool
